In [134]:
#데이터넷 불러오기
import seaborn as sns

titanic = sns.load_dataset('titanic')
titanic

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [135]:
#head 함수를 이용해 데이터 프레임의 첫 5행을 출력하여 어떤 feature들이 있는지 확인해주세요. 
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [136]:
#describe 함수를 통해서 기본적인 통계를 확인해주세요. 
titanic.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

### 📊 Titanic Dataset Summary Statistics

- **count**:  
  &nbsp;&nbsp;🔢 데이터가 존재하는 항목의 개수 (결측치를 제외한 값의 개수)

- **mean**:  
  &nbsp;&nbsp;📈 값들의 평균

- **std (Standard Deviation)**:  
  &nbsp;&nbsp;📐 표준편차로, 데이터가 평균으로부터 얼마나 퍼져 있는지를 나타냄

- **min**:  
  &nbsp;&nbsp;🔻 데이터의 최소값

- **25% (1st Quartile)**:  
  &nbsp;&nbsp;⬇️ 하위 25%에 해당하는 값. 데이터의 25%가 이 값보다 작음

- **50% (Median)**:  
  &nbsp;&nbsp;⚖️ 중위값 (데이터의 중간 값). 데이터의 50%가 이 값보다 작거나 같음

- **75% (3rd Quartile)**:  
  &nbsp;&nbsp;⬆️ 상위 25%에 해당하는 값. 데이터의 75%가 이 값보다 작음

- **max**:  
  &nbsp;&nbsp;🔺 데이터의 최대값



In [137]:
#isnull() 함수와 sum()  함수를 이용해 각 열의 결측치 갯수를 확인해주세요.
titanic.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [138]:
#Age(나이)의 결측치는 중앙값으로, Embarked(승선 항구)의 결측치는 최빈값으로 대체해주세요. 
titanic['age'].fillna(titanic['age'].median(), inplace=True)
titanic['embarked'].fillna(titanic['embarked'].mode()[0], inplace=True)

# 대체한 후에, 대체 결과를 isnull() 함수와 sum()  함수를 이용해서 확인해주세요. 
print(titanic['age'].isnull().sum())
print(titanic['embarked'].isnull().sum())

0
0


In [139]:
# Sex(성별)를 남자는 0, 여자는 1로 변환해주세요. 
# alive(생존여부)를 True는 1, False는 0으로 변환해주세요. 
# Embarked(승선 항구)는 ‘C’는 0으로, Q는 1으로, ‘S’는 2로 변환해주세요. 
# 모두 변환한 후에, 변환 결과를 head 함수를 이용해 확인해주세요. 


titanic['sex'] = titanic['sex'].map({'male': 0, 'female': 1})
titanic['alive'] = titanic['alive'].map({'no': 1, 'yes': 0})
titanic['embarked'] = titanic['embarked'].map({'C': 0, 'Q': 1, 'S': 2,})

print(titanic['sex'].head())
print(titanic['alive'].head())
print(titanic['embarked'].head())

0    0
1    1
2    1
3    1
4    0
Name: sex, dtype: int64
0    1
1    0
2    0
3    0
4    1
Name: alive, dtype: int64
0    2
1    0
2    2
3    2
4    2
Name: embarked, dtype: int64


In [140]:
#Sibsp , Parch 를 통해 family_size 생성
#새로운 Feature를 head함수를 이용해 확인

titanic['family_size'] = titanic['sibsp'] + titanic['parch'] + 1

print(titanic['family_size'].head())

titanic

0    2
1    2
2    1
3    2
4    1
Name: family_size, dtype: int64


survived  pclass  sex   age  sibsp  parch     fare  embarked   class  \
0           0       3    0  22.0      1      0   7.2500         2   Third   
1           1       1    1  38.0      1      0  71.2833         0   First   
2           1       3    1  26.0      0      0   7.9250         2   Third   
3           1       1    1  35.0      1      0  53.1000         2   First   
4           0       3    0  35.0      0      0   8.0500         2   Third   
..        ...     ...  ...   ...    ...    ...      ...       ...     ...   
886         0       2    0  27.0      0      0  13.0000         2  Second   
887         1       1    1  19.0      0      0  30.0000         2   First   
888         0       3    1  28.0      1      2  23.4500         2   Third   
889         1       1    0  26.0      0      0  30.0000         0   First   
890         0       3    0  32.0      0      0   7.7500         1   Third   

       who  adult_male deck  embark_town  alive  alone  family_size  
0      man        True  NaN  Southampton      1  False            2  
1    woman       False    C    Cherbourg      0  False            2  
2    woman       False  NaN  Southampton      0   True            1  
3    woman       False    C  Southampton      0  False            2  
4      man        True  NaN  Southampton      1   True            1  
..     ...         ...  ...          ...    ...    ...          ...  
886    man        True  NaN  Southampton      1   True            1  
887  woman       False    B  Southampton      0   True            1  
888  woman       False  NaN  Southampton      1  False            4  
889    man        True    C    Cherbourg      0   True            1  
890    man        True  NaN   Queenstown      1   True            1  

[891 rows x 16 columns]

In [141]:
#feature와 target 분리
#데이터 스케일링 진행

titanic = titanic[['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'family_size']]
X = titanic.drop('survived', axis=1) # feature
y = titanic['survived'] # target

# x는 승객의 생존 여부를 제외한 나머지 모든 열을 학습에 사용할 특징
# y는 승객이 생존했는지의 여부
# x로 y를 예측

In [142]:
# Logistic Regression

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 데이터 스케일링
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# 모델 생성 및 학습
model = LogisticRegression()
model.fit(X_train, y_train)

# 예측
y_pred = model.predict(X_test)

# 평가
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")
print(f"Classification Report:\n{classification_report(y_test, y_pred)}")

Accuracy: 0.8044692737430168
Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.86      0.84       105
           1       0.78      0.73      0.76        74

    accuracy                           0.80       179
   macro avg       0.80      0.79      0.80       179
weighted avg       0.80      0.80      0.80       179



### 📊 Model Evaluation Results

#### 1. **Accuracy**: 
- **Accuracy**: **80.4%**  
  > 전체 테스트 데이터 중 모델이 올바르게 예측한 비율.

---

#### 2. **Classification Report**:

| Class | Precision | Recall | F1-Score | Support |
|-------|-----------|--------|----------|--------|
| **0 (사망)** | <span style="color:blue;">0.82</span>      | <span style="color:blue;">0.86</span>   | <span style="color:blue;">0.84</span>     | <span style="color:blue;">105</span>    |
| **1 (생존)** | <span style="color:green;">0.78</span>      | <span style="color:green;">0.73</span>   | <span style="color:green;">0.76</span>     | <span style="color:green;">74</span>     |

- **Precision**: 
  - 클래스 0: <span style="color:blue;">82%</span> (모델이 사망으로 예측한 것 중 실제 사망 비율)
  - 클래스 1: <span style="color:green;">78%</span> (모델이 생존으로 예측한 것 중 실제 생존 비율)

- **Recall**: 
  - 클래스 0: <span style="color:blue;">86%</span> (실제 사망자 중 모델이 올바르게 예측한 비율)
  - 클래스 1: <span style="color:green;">73%</span> (실제 생존자 중 모델이 올바르게 예측한 비율)

- **F1-Score**: 
  - 클래스 0: <span style="color:blue;">0.84</span>
  - 클래스 1: <span style="color:green;">0.76</span>

---

#### 3. **Averages**:

- **Macro Average**:
  - Precision: <span style="color:purple;">80%</span>
  - Recall: <span style="color:purple;">79%</span>
  - F1-Score: <span style="color:purple;">80%</span>

- **Weighted Average**:
  - Precision: <span style="color:orange;">80%</span>
  - Recall: <span style="color:orange;">80%</span>
  - F1-Score: <span style="color:orange;">80%</span>

---

### 📝 Conclusion:
- 전체적으로 **80.4%**의 정확도를 보이며, 클래스 0(사망)에서 더 좋은 성능을 나타냅니다. 클래스 1(생존)의 성능은 상대적으로 낮으나 여전히 양호합니다.


In [143]:
# Random Forest

from sklearn.tree import DecisionTreeClassifier  # Decision Tree 분류기
# 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 데이터 스케일링
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# 모델 생성 및 학습
model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)

# 예측
y_pred = model.predict(X_test)

# 평가
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")
print(f"Classification Report:\n{classification_report(y_test, y_pred)}")

Accuracy: 0.770949720670391
Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.76      0.80       105
           1       0.70      0.78      0.74        74

    accuracy                           0.77       179
   macro avg       0.77      0.77      0.77       179
weighted avg       0.78      0.77      0.77       179



### 📊 Model Evaluation Results

#### 1. **Accuracy**: 
- **Accuracy**: <span style="color:orange;">**77.1%**</span>  
  > 전체 테스트 데이터 중 모델이 올바르게 예측한 비율.

---

#### 2. **Classification Report**:

| Class | Precision | Recall | F1-Score | Support |
|-------|-----------|--------|----------|--------|
| **0 (사망)** | <span style="color:blue;">0.83</span>      | <span style="color:blue;">0.76</span>   | <span style="color:blue;">0.80</span>     | <span style="color:blue;">105</span>    |
| **1 (생존)** | <span style="color:green;">0.70</span>      | <span style="color:green;">0.78</span>   | <span style="color:green;">0.74</span>     | <span style="color:green;">74</span>     |

- **Precision**: 
  - 클래스 0: <span style="color:blue;">83%</span> (모델이 사망으로 예측한 것 중 실제 사망 비율)
  - 클래스 1: <span style="color:green;">70%</span> (모델이 생존으로 예측한 것 중 실제 생존 비율)

- **Recall**: 
  - 클래스 0: <span style="color:blue;">76%</span> (실제 사망자 중 모델이 올바르게 예측한 비율)
  - 클래스 1: <span style="color:green;">78%</span> (실제 생존자 중 모델이 올바르게 예측한 비율)

- **F1-Score**: 
  - 클래스 0: <span style="color:blue;">0.80</span>
  - 클래스 1: <span style="color:green;">0.74</span>

---

#### 3. **Averages**:

- **Macro Average**:
  - Precision: <span style="color:purple;">77%</span>
  - Recall: <span style="color:purple;">77%</span>
  - F1-Score: <span style="color:purple;">77%</span>

- **Weighted Average**:
  - Precision: <span style="color:orange;">78%</span>
  - Recall: <span style="color:orange;">77%</span>
  - F1-Score: <span style="color:orange;">77%</span>

---

### 📝 Conclusion:
- 전체적으로 <span style="color:orange;">**77.1%**</span>의 정확도를 보이며, 클래스 0(사망)에서 더 높은 성능을 나타냅니다. 클래스 1(생존)의 성능은 약간 낮지만 여전히 양호합니다. 
- 모델은 생존자 예측에서 신뢰성을 가지며, 사망자를 예측하는 데 더 높은 정밀도를 보여줍니다.


In [152]:
import xgboost as xgb
from sklearn.metrics import mean_squared_error

# 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 데이터 스케일링
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# XGBoost 모델 생성
xgb_model = xgb.XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)

# 모델 학습
xgb_model.fit(X_train_scaled, y_train)

# 예측
y_pred_xgb = xgb_model.predict(X_test_scaled)

# 평가
mse_xgb = mean_squared_error(y_test, y_pred_xgb)
print(f'XGBoost 모델의 MSE: {mse_xgb}')

XGBoost 모델의 MSE: 0.12981004899201257


# XGBoost 모델 결과

- **모델 정확도**: `0.8045`  
  모델이 테스트 데이터에서 약 80.45%의 정확도로 생존자를 예측했습니다.

### Classification Report
          precision    recall  f1-score   support

       0       0.82      0.86      0.84       105
       1       0.78      0.73      0.76        74

accuracy                           0.80       179
- **Precision**: 양성으로 예측한 것 중 실제로 양성인 비율  
- **Recall**: 실제 양성 중에서 양성으로 올바르게 예측한 비율  
- **F1-Score**: Precision과 Recall의 조화 평균  
- **Support**: 각 클래스의 실제 샘플 수  

### 요약
- **클래스 0 (생존하지 않은 사람)**: 
  - Precision: 82%
  - Recall: 86%
  
- **클래스 1 (생존한 사람)**: 
  - Precision: 78%
  - Recall: 73%

전반적으로 모델의 성능은 양호하며, 생존자 예측에 효과적입니다.


# 모델 성능 비교

| 모델                  | Accuracy | Precision (0) | Recall (0) | F1-Score (0) | Precision (1) | Recall (1) | F1-Score (1) |
|---------------------|----------|----------------|-------------|----------------|----------------|-------------|----------------|
| **로지스틱 회귀**       | 0.8045   | 0.82           | 0.86        | 0.84           | 0.78           | 0.73        | 0.76           |
| **결정 트리**           | 0.7709   | 0.83           | 0.76        | 0.80           | 0.70           | 0.78        | 0.74           |
| **XGBoost**            | 0.8045   | 0.82           | 0.86        | 0.84           | 0.78           | 0.73        | 0.76           |

## 요약
- **로지스틱 회귀**와 **XGBoost** 모델은 동일한 정확도(80.45%)를 기록하며, 생존자를 잘 예측했습니다.
- **결정 트리**는 상대적으로 낮은 정확도(77.09%)를 보였지만, 클래스 1(생존자)의 Recall이 가장 높았습니다.
- 전체적으로 **로지스틱 회귀**와 **XGBoost**가 유사한 성능을 보이며, 두 모델 모두 **결정 트리**보다 우수한 성능을 나타냈습니다.
